File: frontend/lib/permissions.ts
/* =========================================================
   SECTION: Imports
   Purpose: External libraries, types, and shared utilities
========================================================= */

import type { UserRole } from "./auth";

/* =========================================================
   SECTION: Constants
   Purpose: Static values, API URLs, role mappings
========================================================= */

export const ALL_MODULE_LABELS = [
  "Dashboard",
  "Users",
  "Branches",
  "CRM",
  "Students",
  "LMS",
  "Finance",
  "HR",
  "AI",
  "Placement",
  "Franchise",
  "Reports",
  "Security",
  "Settings",
] as const;

export const ROLE_MODULE_ACCESS: Record<UserRole, readonly ModuleLabel[]> = {
  super_admin: ALL_MODULE_LABELS,
  branch_admin: ["Dashboard", "Students", "HR", "Finance", "LMS"],
  counsellor: ["Dashboard", "CRM"],
  trainer: ["Dashboard", "LMS", "Students"],
  student: ["Dashboard", "LMS", "AI"],
  parent: ["Dashboard", "Students", "Finance"],
  hr: ["Dashboard", "HR", "Users", "Finance"],
  finance: ["Dashboard", "Finance", "HR"],
  franchise_owner: ["Dashboard", "Franchise", "Branches", "Reports"],
  company_hr: ["Dashboard", "Placement"],
  public: [],
};

export const MODULE_ACTION_ACCESS: Partial<Record<ModuleLabel, readonly UserRole[]>> = {
  Students: ["super_admin", "branch_admin"],
  LMS: ["super_admin", "branch_admin", "trainer"],
  CRM: ["super_admin", "counsellor"],
  Finance: ["super_admin", "branch_admin", "finance"],
  Users: ["super_admin"],
  Security: ["super_admin"],
  Settings: ["super_admin"],
  HR: ["super_admin", "branch_admin", "hr"],
};

/* =========================================================
   SECTION: Types
   Purpose: Shared TypeScript types and interfaces
========================================================= */

export type ModuleLabel = typeof ALL_MODULE_LABELS[number];

/* =========================================================
   SECTION: Helper Functions
   Purpose: Reusable frontend logic
========================================================= */

export function getAllowedModulesForRole(role: UserRole | string | null | undefined): readonly ModuleLabel[] {
  if (!role || !(role in ROLE_MODULE_ACCESS)) return [];
  return ROLE_MODULE_ACCESS[role as UserRole];
}

export function canAccessModule(role: UserRole | string | null | undefined, module: ModuleLabel) {
  return getAllowedModulesForRole(role).includes(module);
}

export function canUseModuleActions(role: UserRole | string | null | undefined, module: ModuleLabel) {
  if (!role) return false;
  const allowedRoles = MODULE_ACTION_ACCESS[module];
  return allowedRoles ? allowedRoles.includes(role as UserRole) : false;
}

/* =========================================================
   SECTION: Exports
   Purpose: Public exports used across the app
========================================================= */

File: backend/app/core/permissions.py
"""
PINESPHERE ERP
Module      : Backend Platform
File        : permissions.py
Purpose     : Provides Permissions backend functionality
Author      : Pinesphere Development Team
Last Updated: Auto Generated
=====================================================
"""

# =====================================================
# SECTION: IMPORTS
# PURPOSE:
# This section loads external libraries, framework tools, and local helpers.
# Keeping imports together makes dependencies easy to review.
# =====================================================

from app.core.roles import UserRole

Permission = str

# =====================================================
# SECTION: CONSTANTS
# PURPOSE:
# This section stores fixed values used by the file.
# Centralizing these values helps avoid repeated magic strings or numbers.
# =====================================================

PERMISSION_ACTIONS = ["view", "create", "edit", "delete", "export", "approve", "analytics"]

ROLE_PERMISSIONS: dict[UserRole, dict[str, list[Permission]]] = {
    UserRole.SUPER_ADMIN: {
        "*": PERMISSION_ACTIONS,
    },
    UserRole.BRANCH_ADMIN: {
        "students": PERMISSION_ACTIONS,
        "staff": PERMISSION_ACTIONS,
        "users": ["view", "create", "edit", "export"],
        "lms": PERMISSION_ACTIONS,
        "attendance": PERMISSION_ACTIONS,
        "finance": PERMISSION_ACTIONS,
    },
    UserRole.COUNSELLOR: {
        "crm": ["view", "create", "edit", "export"],
        "admissions": ["view", "create", "edit", "approve", "export"],
        "leads": ["view", "create", "edit", "export"],
        "students": ["view", "create", "edit"],
    },
    UserRole.TRAINER: {
        "students": ["view", "edit"],
        "lms": ["view", "create", "edit", "analytics"],
        "attendance": ["view", "create", "edit"],
        "batches": ["view", "create", "edit"],
    },
    UserRole.STUDENT: {
        "lms": ["view"],
        "ai_lab": ["view"],
        "internship": ["view"],
        "parent_portal": ["view"],
        "attendance": ["view"],
        "students": ["view"],
    },
    UserRole.PARENT: {
        "students": ["view"],
        "parent_portal": ["view"],
        "notifications": ["view"],
        "attendance": ["view"],
        "finance": ["view"],
        "lms": ["view"],
    },
    UserRole.HR: {
        "hr": PERMISSION_ACTIONS,
        "staff": PERMISSION_ACTIONS,
        "payroll": PERMISSION_ACTIONS,
        "leave": PERMISSION_ACTIONS,
        "performance": PERMISSION_ACTIONS,
        "users": ["view", "create", "edit"],
    },
    UserRole.FINANCE: {
        "finance": PERMISSION_ACTIONS,
        "fees": PERMISSION_ACTIONS,
        "invoicing": PERMISSION_ACTIONS,
        "salary": PERMISSION_ACTIONS,
        "p_and_l": PERMISSION_ACTIONS,
        "students": ["view"],
    },
    UserRole.FRANCHISE_OWNER: {
        "franchise": ["view", "analytics"],
        "branches": ["view", "analytics"],
        "reports": ["view", "export", "analytics"],
    },
    UserRole.COMPANY_HR: {
        "placement": ["view", "create", "edit", "export"],
        "students": ["view"],
    },
    UserRole.PUBLIC: {
        "website": ["view"],
        "admissions": ["view", "create"],
        "bookings": ["view", "create"],
    },
}


# =====================================================
# SECTION: HELPER FUNCTIONS
# PURPOSE:
# This section contains small reusable utilities used by the file.
# Helpers keep repeated logic in one clear place.
# =====================================================

def get_permissions(role: UserRole) -> dict[str, list[Permission]]:
    return ROLE_PERMISSIONS.get(role, {})


def has_permission(role: UserRole, module: str, action: Permission) -> bool:
    permissions = get_permissions(role)
    return action in permissions.get("*", []) or action in permissions.get(module, [])
